Documentation Revision (#25)
ansible/ansible-sign: docs/rundown.rst

Adding a GPG key to AWX or Ansible Automation Controller
=========================================================

In the command line, run the following commands:
``$ gpg --list-keys``
``$ gpg --export --armour <key fingerprint> > my_public_key.asc``

#. In AWX/Automation Controller, click “Credentials" then the "Add" button
#. Give the new credential a meaningful name (for example, "infrastructure team public GPG key")
#. For "Credential Type" select "GPG Public Key"
#. Click "Browse" to navigate to and select the file that you created earlier (``my_public_key.asc``)
#. Finally, click the "Save" button to finish

This credential can now be selected in "Project" settings. Once selected, content verification will automatically take place on future project syncs.

Vist `GPG`_ for more information regarding GPG keys.
For more information regarding generating a GPG key pair, visit the `Red Hat "Enable Sysadmin" blog post`_.

.. _GPG: https://www.gnupg.org/documentation/index.html
.. _Red Hat "Enable Sysadmin" blog post: https://www.redhat.com/sysadmin/creating-gpg-keypairs

How to Access the Ansible Sign CLI Utility
===========================================

> Run the following command to install Ansible-Sign:

Reviewer: ```suggestion
Run the following command to install ``ansible-sign``:
```

.. code-block:: shell

>    :caption: Installing `ansible-sign`

Reviewer: ```suggestion
   :caption: Installing ``ansible-sign``
```

   
   $ pip install ansible-sign

Once it’s installed, run:

.. code-block:: shell

>    :caption: Verify that `ansible-sign` was successfully installed. 

Reviewer: ```suggestion
   :caption: Verify that ``ansible-sign`` was successfully installed. 
```

   
   $ ansible-sign --version

You should see the following output:

.. code-block:: shell

>    :caption: The output of $ ansible-sign --version

Reviewer: ```suggestion
   :caption: The output of ``ansible-sign --version``
```

If the above command produces no output, or one line of output that says that a
"trustdb" was created, then you do not have a secret key in your default

> keyring. Please see the section below titled "Adding a GPG key to AWX or Ansible Automation Controller."

Author: Ah, thank you! This was a remnant when I had from when I had the section about generating the key. I think this reads much better now!

=========================================================

In the command line, run the following commands:
``$ gpg --list-keys``
``$ gpg --export --armour <key fingerprint> > my_public_key.asc``

In AWX/Automation Controller, click “Credentials."
From there, click "Add."
For “Name” name your credential something you will recognize.
For “Credential Type” select  “GPG Public Key."

Click "Browse" to navigate to and select the file that you created earlier. Finally, click the "Save" button to finish.

This credential can now be selected in "Project" settings. Once selected, content verification will automatically take place on future project syncs.

Vist `GPG`_ for more information regarding GPG keys.
For more information regarding generating a GPG key pair, visit the `Red Hat "Enable Sysadmin" blog post`_.

.. _GPG: https://www.gnupg.org/
.. _Red Hat "Enable Sysadmin" blog post: https://www.redhat.com/sysadmin/creating-gpg-keypairs

How to Access the Ansible Sign CLI Utility:
===========================================

Run the following command to install Ansible-Sign:
``$ pip install ansible-sign ``

Once it’s installed, run:

> ``$ ansible-sign --help`` 

Reviewer: I wonder if instead of `--help` we give the output of `--version` or something, if the goal here is just a self-check that the script installed correctly.

The output of `--version` is likely to be more stable (well, the version will change of course, but the format isn't likely to). This eliminates needing to update this document every time a new flag or top-level subcommand is added to the utility.

Author: I agree! That definitely makes sense. I have made that change!

If the above command produces no output, or one line of output that says that a
"trustdb" was created, then you do not have a secret key in your default
keyring.Please see the section below titled "Adding a GPG key to AWX or Ansible Automation Controller."

If it produces output other than that, then you have a valid secret key
and are ready to move on to "ow to Access the Ansible Sign CLI Utility."

Adding a GPG key to AWX or Ansible Automation Controller:
=========================================================

In the command line, run the following commands:
``$ gpg --list-keys``
``$ gpg --export --armour <key fingerprint> > my_public_key.asc``

In AWX/Automation Controller, click “Credentials."
From there, click “Add”.
For “Name” name your credential something you will recognize.
For “Credential Type” select  “GPG Public Key."

Click "Browse" to navigate to and select the file that you created earlier. Great work! You're ready to start verifying signed project content. 

Additional documentation:

> .. _GPG: https://www.gnupg.org/

Author: @relrod I moved these links so they make more sense contextually, but now they aren't showing up. I can't seem to figure out what changed to cause that.

.. _GPG: https://www.gnupg.org/
.. _Red Hat "Enable Sysadmin" blog post: https://www.redhat.com/sysadmin/creating-gpg-keypairs

How to Access the Ansible Sign CLI Utility:
===========================================

In the directory in which you would like Ansible Sign to be installed, run:
``$ pip install ansible-sign .``

Once it’s installed, run:
``$ ansible-sign --help`` 

You should see the following output:

.. code-block:: shell
   :caption: The output of $ ansible-sign --help

   ansible-sign [-h] [--version] [--debug] {project} ...

   Signing and validation for Ansible content

   positional arguments:
   {project}
   project   Act on an Ansible project directory

   options:
   -h, --help  show this help message and exit
   --version   show program's version number and exit
   --debug     Print a bunch of debug info
   --nocolor   Disable color output

> Congratulations! You have successfully installed Ansible-Sign!

Reviewer: ```suggestion
Congratulations! You have successfully installed ``ansible-sign``!
```

.. _GPG: https://www.gnupg.org/
.. _Red Hat "Enable Sysadmin" blog post: https://www.redhat.com/sysadmin/creating-gpg-keypairs

How to Access the Ansible Sign CLI Utility:
===========================================
Clone ansible-sign as an entirely new repo (This is typically placed within an existing AWX repo).

``$ git clone git@github.com:ansible/ansible-sign``

There are certainly alternative methods for closing a repo, absolutely use whichever method you prefer. (You might use github's https cloning instead of the git@ syntax, which is SSH.)

Then once you have that cloned, you can:
``$ cd ansible-sign``

Then:

> ``$ pip install -e .``

Reviewer: This is a good developer workflow for people who want to work on ansible-sign for any reason. For end users, we probably want them to install it from PyPI (or from dnf if they are using RHEL/AAP?).

So all of these lines can effectively just become:
> Run `pip install ansible-sign`.

... for end users.

Author: Thank you so much! I will revise this bit accordingly!

Author: I have this saved in my notes for if we want to make a "How to contribute" page later. I am updating as per your instructions.

That command will install it. Once it’s installed, run:
``$ ansible-sign --help``

You should see the following output: 
.. code-block:: shell
usage: ansible-sign [-h] [--version] [--debug] {project} ...

Signing and validation for Ansible content

positional arguments:
  {project}
    project   Act on an Ansible project directory

options:
  -h, --help  show this help message and exit
  --version   show program's version number and exit
  --debug     Print a bunch of debug info

>   --nocolor   Disable color output

Reviewer: This will probably have to be indented to render correctly.

How to Access the Ansible Sign CLI Utility:
===========================================

> In the directory in which you would like Ansible Sign to be installed, run:

Author: Thank you, I have removed this portion!

> ``$ pip install ansible-sign .``

Author: Fixed!

   take the project root directory as their last argument, thus we will simply
   use ``.`` to indicate the current Working Directory.

Adding a GPG key to AWX or Ansible Automation Controller:
=========================================================

In the command line, run the following commands:
``$ gpg --list-keys``
``$ gpg --export --armour <key fingerprint> > my_public_key.asc``

In AWX/Automation Controller, click “Credentials."
From there, click “Add”.
For “Name” name your credential something you will recognize.
For “Credential Type” select  “GPG Public Key."

> Click "Browse" to navigate to and select the file that you created earlier. Great work! You're ready to start verifying signed project content. 

Reviewer: Might be obvious but for completion might as well mention "click the save button" to finish.

Also, maybe the end should be something like "This credential can now be selected in Project settings. Once selected, content verification will automatically take place on future project syncs."

This can be reworded in whatever way makes sense.

   root of your project. ``ansible-sign project`` commands, as a rule, always
   take the project root directory as their last argument, thus we will simply
   use ``.`` to indicate the current Working Directory.
Generating a GPG Key
===================

Full documentation:

> https://docs.github.com/en/authentication/managing-commit-signature-verification/generating-a-new-gpg-key

Reviewer: This might be okay to link to, but I'd like someone else to sign off on that before we do it. There's a RH blog post about making keys and I link to it at the very start of this file.

Follow the link above to download the GPG Command Line Tools.

Once you have the GPG Command line tools, follow these steps to generate your key.

``$ gpg --full-generate-key`` is the command to begin the key generation process.

For most cases, just press enter to accept the default settings. 
For the expiration date, no expiration date is the default. 
Once all of that is verified, enter your user ID information. This should be your GitHub email for the email address.

``$ gpg --list-secret-keys --keyid-format=long`` will list the key that has just been created in long form. 

Congratulations! You’re ready to add your key to the Ansible UI.

> How to add GPG key to Ansible UI:

Reviewer: ```suggestion
Adding a GPG key to AWX or Ansible Automation Controller:
```

================================

In the command line, run the following commands:
``$ gpg --list-keys``
``$ gpg --export --armour <key fingerprint>``

Copy  everything from:
``-----BEGIN PGP PUBLIC KEY BLOCK-----``
to 

> ``-----END PGP PUBLIC KEY BLOCK-----``

Reviewer: It is probably better to have people upload it as a file to prevent copy-paste mistakes.

`gpg --export --armour <key fingerprint> > my_public_key.asc` and then in AWX instead of pasting, click browse and navigate to that file and it will fill it in for you.

> In the Ansible UI, click “Credentials."

Reviewer: ```suggestion
In AWX/Automation Controller, click “Credentials."
```

From there, click “Add”.
For “Name” name your credential something you will recognize.

> For “Credential Type” select  “GitHub Personal Access Token."

Reviewer: ```suggestion
For “Credential Type” select  “GPG Public Key."
```

> Paste what you copied from your command line earlier into the “Type Details” box. Great work! You're ready to start signing content. 

Reviewer: ```suggestion
Paste what you copied from your command line earlier into the “Type Details” box. Great work! You're ready to start verifying signed project content. 
```

but see note above, it's probably better to have them upload the file :slightly_smiling_face:

   :caption: Generating a checksum manifest file and signing it

   $ ansible-sign project gpg-sign .
[OK   ] GPG signing successful!
[NOTE ] Checksum manifest: ./.ansible-sign/sha256sum.txt

> [NOTE ] GPG summary: signature created

Reviewer: These can't be dedented because they won't be in the code bock anymore, rst is indentation dependent.

   :caption: Verifying our sample project

   $ ansible-sign project gpg-verify .

>    [OK   ] GPG signature verification succeeded.

Reviewer: Is this missing a line? There should be a line showing for both verifications if I didn't break something in ansible-sign.

Author: I think it might be the way mine is set up- It is giving me an error message. Perhaps we could check on this together next week? Alternatively, if you wouldn't mind giving me the expected output I can add it here. Whichever you prefer!

Reviewer: Happy to work with you on figuring out where the error is coming from.

     - Signing process failure
     - * The signer's private key was not found in the GPG keyring
       * The wrong GnuPG home directory or keyring file was specified

Other Helpful Commands:
======================
``$ time ansible-sign project gpg-sign .``

> This command shows how long validation took.

Reviewer: I don't know if this will be too useful to people. I ran it on a repo with ~4300 files and took under 0.4s for both signing and verifying. I think generally this shouldn't be an issue.

Additional Resources:
====================

.. _AWX project sync via action plugin: https://www.youtube.com/watch?v=srajyQpzkmI 

> .. _Ansible-sign CLI: https://www.youtube.com/watch?v=fyb0OxejYnk

Reviewer: I don't think I would include these. They were meant to be internal demos and also some things have changed since recording them.
